### ISSUE TITLE
Out-Null is not expensive to pipe to

### ISSUE THREAD

**Issue by powercode** [MicrosoftDocs/PowerShell-Docs#8333]
This was fixed quite a while ago.

See src\System.Management.Automation\engine\runtime\Operations\MiscOps.cs:468-475

```csharp
                var cmdletInfo = commandProcessor?.CommandInfo as CmdletInfo;
                if (cmdletInfo?.ImplementingType == typeof(OutNullCommand))
                {
                    var commandsCount = pipelineProcessor.Commands.Count;
                    if (commandsCount == 1)
                    {
                        // Out-Null is the only command, bail without running anything
                        return;
                    }

                    // Out-Null is the last command, rewrite command before Out-Null to a null pipe, but
                    // only if it didn't redirect anything, e.g. `Get-Stuff > o.txt | Out-Null`
                    var nextToLastCommand = pipelineProcessor.Commands[commandsCount - 2];
                    if (!nextToLastCommand.CommandRuntime.OutputPipe.IsRedirected)
                    {
                        pipelineProcessor.Commands.RemoveAt(commandsCount - 1);
                        commandProcessor = nextToLastCommand;
                        nextToLastCommand.CommandRuntime.OutputPipe = new Pipe { NullPipe = true };
                    }
                }
```

A null pipe is created if piped to out-null. It is roughly 10% slower than assigning to `$null`, but in most cases not even noticeable.

```powershell
PS > measure-command { 1..1000000 | foreach{$_} | out-null} | ft
```
```text
Days Hours Minutes Seconds Milliseconds
---- ----- ------- ------- ------------
0    0     0       3       701
```
```powershell
PS > measure-command { $null = 1..1000000 | foreach{$_}} | ft
```
```text
Days Hours Minutes Seconds Milliseconds
---- ----- ------- ------- ------------
0    0     0       3       51
```

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: e911b722-86ce-7553-e9ab-33927e126b46
* Version Independent ID: 9818d91b-8c5f-337a-f5e1-7a40abd5fd1a
* Content: [PowerShell scripting performance considerations - PowerShell](https://docs.microsoft.com/en-us/powershell/scripting/dev-cross-plat/performance/script-authoring-considerations?view=powershell-7.2)
* Content Source: [reference/docs-conceptual/dev-cross-plat/performance/script-authoring-considerations.md](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/docs-conceptual/dev-cross-plat/performance/script-authoring-considerations.md)
* Product: **powershell**
* Technology: **powershell-conceptual**
* GitHub Login: @sdwheeler
* Microsoft Alias: **sewhee**

**Comment by JamesWTruher** [comment]
here's a counterpoint to your measurement:

```powershell
PS> $d = get-date ; measure-command { for($i = 0; $i -lt 1mb;$i++) { $null=$d } }|ft

Days Hours Minutes Seconds Milliseconds
---- ----- ------- ------- ------------
0    0     0       0       859

PS> $d = get-date ; measure-command { for($i = 0; $i -lt 1mb;$i++) { $d|out-null } }|ft

Days Hours Minutes Seconds Milliseconds
---- ----- ------- ------- ------------
0    0     0       4       131

PS> $d = get-date ; measure-command { $null = for($i = 0; $i -lt 1mb;$i++) { $d } }|ft

Days Hours Minutes Seconds Milliseconds
---- ----- ------- ------- ------------
0    0     0       0       925

PS> $d = get-date ; measure-command { $(for($i = 0; $i -lt 1mb;$i++) { $d })|out-null }|ft

Days Hours Minutes Seconds Milliseconds
---- ----- ------- ------- ------------
0    0     0       0       923

```
I think it rather depends on the location of the pipe to `out-null` and how many times it's done. Certainly, one can write pathological code (as I have done above) which _seems_ to be much slower but comparing apples to apples it's not.

**Comment by daxian-dbw** [comment]
Yeah, the optimization doesn't kick in when `Out-Null` is the only command in the pipeline:
```
                    var commandsCount = pipelineProcessor.Commands.Count;
                    if (commandsCount == 1)
                    {
                        // Out-Null is the only command, bail without running anything
                        return;
                    }
```
Also, the optimization change is not included in Windows PowerShell 5.1.

**Comment by JamesWTruher** [comment]
also, as a pattern, i would suggest that using `| out-null` is not one we should promulgate
